### PR TITLE
feat(sms): Add an allow list of non-rate limited phone numbers for testing.

### LINKED ISSUE #217: Allow list of test phone numbers that will not be rate limited.
Right now the content-server has a bunch of SMS related functional tests that aren't run on "prod-like" environments because they send actual SMSs. Instead of using randomly generated phone numbers, I tried using test phone numbers provided by phone companies, but was soon rate limited.

We should allow SMS messages to be sent to these special phone numbers w/o being rate limited, perhaps keeping a list of numbers or regexps that aren't rate limited - sort of like we do with email addresses.

Ref mozilla/fxa-content-server#5689
Replaces https://github.com/mozilla/fxa-auth-server/issues/2220

### PR DESCRIPTION
**Pull request by shane-tomlinson**
fixes #217

@rfk - r?

### CODE REVIEW

File: lib/settings/allowed_phone_numbers.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      super('allowedPhoneNumbers')
+      this.setAll(phoneNumbers)
+    }
+
+    isAllowed(phoneNumber) {
+      return phoneNumber in this.phoneNumbers
+    }
+
+    setAll(phoneNumbers) {
+      this.phoneNumbers = {}
```
Comment (shane-tomlinson) [inline lib/settings/allowed_phone_numbers.js]: I wasn't sure if a JavaScript Set is allowed in the customs-server, seems like it'd be a natural fit.

Comment (philbooth) [reply lib/settings/allowed_phone_numbers.js]: I say do it, good shout!

File: lib/server.js [hunk 2]
```diff
@@ -129,13 +131,32 @@ module.exports = function createServer(config, log) {
     return P.all(promises)
   }
 
-  function allowWhitelisted (result, ip, email) {
-    // Regardless of any preceding checks, there are some IPs and emails
-    // that we won't block. These are typically for Mozilla QA purposes.
-    // They should be checked after everything else so as not to pay the
-    // overhead of checking the many requests that are *not* QA-related.
+  function isIpAllowed(ip) {
```
Comment (vbudhram) [inline lib/server.js]: With the introduction of these, wonder if we should we pull configs and logic into a common helper class and export `allowWhitelisted`?

File: lib/server.js [hunk 3]
```diff
@@ -262,7 +283,7 @@ module.exports = function createServer(config, log) {
         )
         .then(
           function (result) {
-            allowWhitelisted(result, ip, email)
+            allowWhitelisted(result, ip, email, phoneNumber)
```
Comment (vbudhram) [inline lib/server.js]: Can probably add a `getAllowSMS` endpoint similar to [these](https://github.com/mozilla/fxa-customs-server/blob/78226031d44e79064ac50a79c696553ff8540155/lib/server.js#L609).

File: lib/server.js (hunk 2 above)
Comment (shane-tomlinson) [reply lib/server.js]: Does `isAllowed` on `allowedEmailDomains` and `allowedPhoneNumbers` fit that?

File: lib/server.js (hunk 3 above)
Comment (shane-tomlinson) [reply lib/server.js]: On it.

**Comment by shane-tomlinson** [comment]
> couldn't find anything on Nexmo's site, but can we send sms to certain numbers for free?

We switched to Amazon a few months back, I couldn't find anything on their docs either.

File: lib/server.js [hunk 4]
```diff
@@ -129,13 +131,20 @@ module.exports = function createServer(config, log) {
...
+
+  function allowWhitelisted (result, ip, email, phoneNumber) {
+    // Regardless of any preceding checks, there are some IPs, emails,
+    // and phone numbers that we won't block. These are typically for
+    // Mozilla QA purposes. They should be checked after everything
+    // else so as not to pay the overhead of checking the many requests
+    // that are *not* QA-related.
     if (result.block || result.suspect) {
-      if (ip in allowedIPs.ips || (email && allowedEmailDomains.isAllowed(email))) {
+      if (isAllowed(ip, email, phoneNumber)) {
```
Comment (rfk) [inline lib/server.js]: Nice cleanup 👍

File: scripts/config-list.js [hunk 5]
```diff
@@ -41,6 +41,7 @@ var output = {
   limits: {},
   allowedIPs: [],
   allowedEmailDomains: [],
+  allowedPhoneNumbers: [],
```
Comment (rfk) [inline scripts/config-list.js]: IIUC, I think we also need to add logic to actually fetch `allowedPhoneNumbers` from memcache, similar to `mc.getAsync('allowedEmailDomains')`?

Comment (shane-tomlinson) [reply scripts/config-list.js]: [Added](https://github.com/mozilla/fxa-customs-server/pull/218/files#diff-51918078d313a138e042691cb119ed4dR79) - mind having a look?

**Pull request merged**